### PR TITLE
[SPARK-13569][STREAMING][KAFKA] pattern based topic subscription

### PR DESCRIPTION
**Pull request by koeninger**
## What changes were proposed in this pull request?

Allow for kafka topic subscriptions based on a regex pattern.
## How was this patch tested?

Unit tests, manual tests

### CODE REVIEW

**Comment by koeninger** [comment]
@tdas @zsxwing This should be the last ConsumerStrategy implementation to have basic parity with what's offered by the kafka consumer, anything else should probably be handled by user subclasses.

If the KAFKA-3370 workaround stuff isn't clear... the basic issue is that you have to poll in order to get partition assignments before setting a position... but if you poll with auto offset none, it will throw an exception because you don't have a position yet :)

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala [hunk 1]
```diff
@@ -79,8 +81,71 @@ private case class Subscribe[K, V](
...
+      val aor = kafkaParams.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG)
+      val shouldSuppress = aor != null && aor.asInstanceOf[String].toUpperCase == "NONE"
+      try {
+        consumer.poll(0)
+      } catch {
+        case x: NoOffsetForPartitionException if shouldSuppress =>
+          // silence exception
+      }
+      toSeek.asScala.foreach { case (topicPartition, offset) =>
+          consumer.seek(topicPartition, offset)
```
Comment (jaceklaskowski) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: 4 chars for indent?

Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: Foreach is a scope, case is a nested scope.

**Comment by koeninger** [comment]
ping @tdas @zsxwing these poll fixes need to go in for a release / release candidate, even if SubscribePattern doesn't make it for some reason.

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala [hunk 2]
```diff
@@ -79,8 +81,71 @@ private case class Subscribe[K, V](
...
     consumer.subscribe(topics)
-    if (currentOffsets.isEmpty) {
-      offsets.asScala.foreach { case (topicPartition, offset) =>
+    val toSeek = if (currentOffsets.isEmpty) {
+      offsets
+    } else {
+      currentOffsets
+    }
+    if (!toSeek.isEmpty) {
+      // work around KAFKA-3370 when reset is none
```
Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: can you explain the problem and the work around?

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala [hunk 3]
```diff
@@ -79,8 +81,71 @@ private case class Subscribe[K, V](
   def onStart(currentOffsets: ju.Map[TopicPartition, jl.Long]): Consumer[K, V] = {
     val consumer = new KafkaConsumer[K, V](kafkaParams)
     consumer.subscribe(topics)
-    if (currentOffsets.isEmpty) {
-      offsets.asScala.foreach { case (topicPartition, offset) =>
+    val toSeek = if (currentOffsets.isEmpty) {
+      offsets
+    } else {
+      currentOffsets
```
Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: This is a change from the semantics with current offsets. earlier logic was: if there is current offsets (e.g. restart from checkpoint), dont seek. This was fine because the DStream new the starting offsets of the next batch, so setting the current position of the consumer was not needed. 

Why this change now?

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala [hunk 4]
```diff
@@ -79,8 +81,71 @@ private case class Subscribe[K, V](
...
+    }
+    if (!toSeek.isEmpty) {
```
Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: nit: toSeek.nonEmpty

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala [hunk 5]
```diff
@@ -79,8 +81,71 @@ private case class Subscribe[K, V](
...
+      // work around KAFKA-3370 when reset is none
+      val aor = kafkaParams.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG)
+      val shouldSuppress = aor != null && aor.asInstanceOf[String].toUpperCase == "NONE"
+      try {
+        consumer.poll(0)
+      } catch {
+        case x: NoOffsetForPartitionException if shouldSuppress =>
+          // silence exception
```
Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: I dont think we should silence it, just log it as WARN.

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala [hunk 6]
```diff
@@ -79,8 +81,71 @@ private case class Subscribe[K, V](
...
+      currentOffsets
+    }
+    if (!toSeek.isEmpty) {
+      // work around KAFKA-3370 when reset is none
+      val aor = kafkaParams.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG)
+      val shouldSuppress = aor != null && aor.asInstanceOf[String].toUpperCase == "NONE"
+      try {
+        consumer.poll(0)
+      } catch {
+        case x: NoOffsetForPartitionException if shouldSuppress =>
```
Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: same here. log as warn.

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala (hunk 4 above)
Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: Java maps don't have nonempty, right?

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala (hunk 3 above)
Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: The consumer needs to be in a state that poll() is safe to call on it.  With auto offset reset none, a consumer with no positions will throw if poll is called.  You're right that the driver knows the recovered positions, this is more about working around consumer state.

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala (hunk 2 above)
Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: Sure, will add comment in the code once I'm back at my workstation.

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala (hunk 5 above)
Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: Can do... Was a little gun shy about extending logging trait in these classes.

File: external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala (hunk 3 above)
Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: but you are calling poll before the seek? why is that?

Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: Because without poll, for subscriptions you don't have partition assignments.  It's an interesting consumer api, to say the least.

Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: guess it wasnt designed for rapidly seeking back and forth. so to confirm my understanding, these are the following calls in Subscribe

- poll - to get the initial assignments. handles none case of no committed offset
- seek - set the position to whatever is known from user-provided offsets or current offsets, so that later regular polls in the DStream can proceed with as little difficulty as possible.

right?

Comment (tdas) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: I saw the comments below. That's confirmed. Thanks for adding the comments!

Comment (koeninger) [inline external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/ConsumerStrategy.scala]: Basically, yeah.

**Comment by tdas** [comment]
I am going to merge this PR in the interest of RC2. But I would like to have more tests for testing the conditions that led to the poll+seek to be added. These subtle behaviors should not go untested.
